middelware.go

```diff
@@ -309,3 +314,180 @@ func ServiceLinkMiddleware() Middleware {
+		q := 1.0
+		if len(parts) > 1 {
+			trimmed := strings.Trim(parts[1], " \t")
+			if strings.HasPrefix(trimmed, "q=") {
+				q, _ = strconv.ParseFloat(trimmed[2:], 64)
+			}
+		}
+
+		// Prefer the first one if there is a tie.
+		if q > bestQ || (q == bestQ && name == allowed[0]) {
+			bestQ = q
+			best = name
+		}
+	}
+
+	return best
+}
+
+type contentEncodingWriter struct {
+	gin.ResponseWriter
+	status      int
+	encoding    string
+	buf         *bytes.Buffer
+	writer      io.Writer
+	minSize     int
+	gzPool      *sync.Pool
+	brPool      *sync.Pool
+	wroteHeader bool
+}
+
+func (w *contentEncodingWriter) Write(data []byte) (int, error) {
+	if w.writer != nil {
+		// We are writing compressed data.
+		return w.writer.Write(data)
+	}
+
+	// Buffer the data until we can decide whether to compress it or not.
+	w.buf.Write(data)
+
+	cl, _ := strconv.Atoi(w.Header().Get("Content-Length"))
+	if cl >= w.minSize || w.buf.Len() >= w.minSize {
+		// We reached out minimum compression size. Set the writer, write the buffer
+		// and make sure to set the correct headers.
+		switch w.encoding {
+		case "gzip":
+			gz := w.gzPool.Get().(*gzip.Writer)
+			gz.Reset(w.ResponseWriter)
+			w.writer = gz
+		case "br":
+			br := w.brPool.Get().(*brotli.Writer)
+			br.Reset(w.ResponseWriter)
+			w.writer = br
+		}
+		w.Header().Set("Content-Encoding", w.encoding)
+		w.Header().Set("Vary", "Accept-Encoding")
+		w.ResponseWriter.WriteHeader(w.status)
+		w.wroteHeader = true
+		bufData := w.buf.Bytes()
+		w.buf.Reset()
+		return w.writer.Write(bufData)
+	}
+
+	// Not sure yet whether this should be compressed.
+	return len(data), nil
+}
+
+func (w *contentEncodingWriter) WriteHeader(code int) {
+	w.Header().Del("Content-Length")
+	w.status = code
+}
+
+func (w *contentEncodingWriter) Close() {
+	if !w.wroteHeader {
+		w.ResponseWriter.WriteHeader(w.status)
+	}
+
+	if w.buf.Len() > 0 {
+		w.ResponseWriter.Write(w.buf.Bytes())
+	}
+
+	if w.writer != nil {
+		if wc, ok := w.writer.(io.WriteCloser); ok {
+			wc.Close()
+		}
+	}
+}
+
+// ContentEncodingMiddleware uses content negotiation with the client to pick
+// an appropriate encoding (compression) method and transparently encodes
+// the response. Supports GZip and Brotli.
+func ContentEncodingMiddleware() Middleware {
+	// Use pools to reduce allocations. We use a byte buffer to temporarily store
+	// some of each response in order to determine whether compression should
+	// be applied. The others are just re-using the GZip and Brotli compressors.
+	bufPool := sync.Pool{
+		New: func() interface{} {
+			return new(bytes.Buffer)
+		},
+	}
+
+	gzPool := sync.Pool{
+		New: func() interface{} {
+			return gzip.NewWriter(ioutil.Discard)
+		},
+	}
+
+	brPool := sync.Pool{
+		New: func() interface{} {
+			return brotli.NewWriter(ioutil.Discard)
+		},
+	}
+
+	return func(c *gin.Context) {
+		if ext := path.Ext(c.Request.URL.Path); ext == ".gif" || ext == ".png" || ext == ".jpg" || ext == ".jpeg" || ext == ".zip" {
```

Comment:
This looks like it should be a list of "excluded path extensions" that should be be extracted as a magic array. I would also suggest adding `.gz` and `.bz2` to keep from trying to double-compress things.